source/agora/cli/multi/main.d

```diff
@@ -0,0 +1,280 @@
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.cli.multi.main;
+
+import agora.api.FullNode;
+
+import agora.common.Config;
+import agora.node.Node;
+import agora.utils.Log;
+
+import vibe.core.core;
+import vibe.http.router;
+import vibe.http.server;
+import vibe.web.rest;
+
+import ocean.util.log.Logger;
+
+import std.file;
+import std.format;
+import std.path;
+import std.range;
+import std.stdio;
+import std.string;
+
+import core.thread;
+
+mixin AddLogger!();
+
+/// Default port for nodes when forwarding ports
+private immutable ushort NODE_BASE_PORT = 4000;
+
+/// Admin port for nodes when forwarding ports
+private immutable ushort ADMIN_BASE_PORT = 5000;
+
+/*******************************************************************************
+
+    Application entry point
+
+*******************************************************************************/
+
+version (unittest) {}
+else
+private int main (string[] args)
+{
+    // Validate parameters
+    if (args.length < 2)
+    {
+        printDefaultHelp();
+        return 1;
+    }
+
+    string[] files;
+    // If the root path is not set,
+    if (extension(args[1]).toLower == ".yaml")
+        files = args[1..$];
+    else
+    // If the root path is not set,
+        foreach (idx; 2..args.length)
+            files ~= buildNormalizedPath(args[1], args[idx]);
+
+    // Reads all configurations from the file.
+    Config[] configs;
+    foreach (idx, ref file; files)
+    {
+        try
+        {
+            // Read
+            auto config = readFromFile(file);
+
+            // Convert data to suit the environment.
+            config.node.address = "127.0.0.1";
+            config.node.port = cast(ushort)(NODE_BASE_PORT + idx);
+            config.node.data_dir = buildNormalizedPath(dirName(file), config.node.data_dir);
+            config.admin.port = cast(ushort)(ADMIN_BASE_PORT + idx);
+            config.logging.log_level = LogLevel.Error;
+
+            string[] coverted_network;
+            foreach (address; config.network)
+                coverted_network ~= replaceAddress(address, "127.0.0.1");
+
+            Config coverted_config =
+            {
+                banman : config.banman,
+                node : config.node,
+                network : cast(immutable string[])coverted_network,
+                dns_seeds : config.dns_seeds,
+                quorum : config.quorum,
+                logging: config.logging,
+                admin: config.admin,
+            };
+
+            // Insert
+            configs ~= coverted_config;
+        }
+        catch (Exception ex)
+        {
+            writefln("Failed to parse config file '%s'. Error: %s",
+                file, ex.message);
+            return 1;
+        }
+    }
+
+    NodeLauncher[] launchers;
+    foreach (ref config; configs)
```

Comment:
```suggestion
    foreach (const ref config; configs)
```

If it's not being modified, we should use `const`. And `ref` is there to avoid copying by value, which can be expensive.